### PR TITLE
GH-338: [Java][CI] Upgrade Arrow Java version for Java Cookbooks

### PR DESCRIPTION
**Pull request by davisusanibar**
To closes https://github.com/apache/arrow-cookbook/issues/338

### CODE REVIEW

**Comment by davisusanibar** [comment]
Need to review https://nightlies.apache.org/arrow/java/org/apache/arrow/ for new Java modules added:
- bom
- maven
- netty patch

**Comment by raulcd** [comment]
Thanks @davisusanibar , as there is a small script to update versions I would run the automated one and do it on a single PR. See my comment here: https://github.com/apache/arrow-cookbook/pull/343

**Comment by davisusanibar** [comment]
Is needed to review Arrow Java Dataset on Linuz environment.

- **MacOS: OK**
- **Linux: NOK**

````
2024-02-07T16:15:45.7564602Z Document: dataset
2024-02-07T16:15:45.7565106Z -----------------
2024-02-07T16:15:49.6220918Z SLF4J(W): No SLF4J providers were found.
2024-02-07T16:15:49.6646043Z SLF4J(W): Defaulting to no-operation (NOP) logger implementation
2024-02-07T16:15:49.6647689Z SLF4J(W): See https://www.slf4j.org/codes.html#noProviders for further details.
2024-02-07T16:15:49.6649325Z SLF4J(W): Class path contains SLF4J bindings targeting slf4j-api versions 1.7.x or earlier.
2024-02-07T16:15:49.6651710Z SLF4J(W): Ignoring binding found at [jar:file:/home/runner/.m2/repository/org/slf4j/slf4j-jdk14/1.7.30/slf4j-jdk14-1.7.30.jar!/org/slf4j/impl/StaticLoggerBinder.class]
2024-02-07T16:15:49.6654062Z SLF4J(W): See https://www.slf4j.org/codes.html#ignoredBindings for an explanation.
2024-02-07T16:15:50.0969022Z Exception java.lang.UnsatisfiedLinkError: /tmp/jnilib-5670360797150962767.tmp: /tmp/jnilib-5670360797150962767.tmp: undefined symbol: _ZTIN6google8protobuf7MessageE
2024-02-07T16:15:50.0970774Z       at NativeLibraries.load (Native Method)
2024-02-07T16:15:50.0977239Z       at NativeLibraries$NativeLibraryImpl.open (NativeLibraries.java:395)
2024-02-07T16:15:50.0978296Z       at NativeLibraries.loadLibrary (NativeLibraries.java:234)
2024-02-07T16:15:50.0979275Z       at NativeLibraries.loadLibrary (NativeLibraries.java:176)
2024-02-07T16:15:50.0980137Z       at ClassLoader.loadLibrary (ClassLoader.java:2394)
2024-02-07T16:15:50.1068329Z       at Runtime.load0 (Runtime.java:785)
2024-02-07T16:15:50.1070185Z       at System.load (System.java:1979)
2024-02-07T16:15:50.1071985Z       at JniLoader.load (JniLoader.java:92)
2024-02-07T16:15:50.1073983Z       at JniLoader.loadRemaining (JniLoader.java:75)
2024-02-07T16:15:50.1076186Z       at JniLoader.ensureLoaded (JniLoader.java:61)
2024-02-07T16:15:50.1077401Z       at NativeMemoryPool.getDefault (NativeMemoryPool.java:34)
2024-02-07T16:15:50.1078292Z       at (#13:3)
2024-02-07T16:15:50.4281061Z **********************************************************************
2024-02-07T16:15:50.4291663Z File "dataset.rst", line 65, in default
````

File: java/source/conf.py [hunk 1]
```diff
@@ -38,9 +38,9 @@
 author = 'The Apache Software Foundation'
 arrow_nightly=os.getenv("ARROW_NIGHTLY")
 if arrow_nightly and arrow_nightly != '0':
-    version = "15.0.0-SNAPSHOT"
+    version = "15.0.0"
```
Comment (vibhatha) [inline java/source/conf.py:41]:
```suggestion
    version = "16.0.0-SNAPSHOT"
```

File: java/ext/javadoctest.py [hunk 2]
```diff
@@ -47,6 +47,10 @@ def compile(
     ) -> Any:
         # go to project that contains all your arrow maven dependencies
         path_arrow_project = pathlib.Path(__file__).parent.parent / "source" / "demo"
+        # delete classpath java dependencies files used
```
Comment (vibhatha) [inline java/ext/javadoctest.py:50]: What is the main reason to add this for Arrow Java version upgrade?

Comment (davisusanibar) [reply java/ext/javadoctest.py:50]: Hi @vibhatha this change was for testing purpose, let me update with final changes

**Comment by vibhatha** [comment]
Seems like all the CIs are passing. Does it mean that we have resolved the protobuf related problem?

**Comment by davisusanibar** [comment]
> Seems like all the CIs are passing. Does it mean that we have resolved the protobuf related problem?

You're right, it is. PR related: https://github.com/apache/arrow/pull/40015

**Comment by davisusanibar** [comment]
I would appreciate a review from you @lidavidm.

**Review by lidavidm — APPROVED** [review]

**Pull request merged**